### ISSUE TITLE
Slow cram performance due to hflush

### ISSUE THREAD

**Issue by jkbonfield** [samtools/htslib#883]
Ancient history for @jmarshall to maybe wrack his brain on.

Commit 6416ebf2 made the CRAM code use hfile instead of FILE pointers.  In doing so, an fflush call was replaced by hflush.

Naively this seems like a sensible option and a simple one to one replacement.  However, hflush does an fdatasync, which turns it into a synchronous write.  The performance hit appears to be OS specific, but on my desktop running Ubuntu Bionic:

With no flush:

```
time samtools view -o /tmp/v4.cram ~/scratch/data/v4_demo/novaseq.bam

real	0m4.312s
user	0m3.798s
sys	0m0.200s
```

With fflush:
```
time samtools view -o /tmp/v4.cram ~/scratch/data/v4_demo/novaseq.bam

real	0m4.263s
user	0m3.759s
sys	0m0.157s
```

With hflush
```
time samtools view -o /tmp/v4.cram ~/scratch/data/v4_demo/novaseq.bam

real	0m8.631s
user	0m3.903s
sys	0m0.157s
```

Basically fflush is pretty much free, but we're half speed with hflush.  I don't think this change was ever intended to operate this way.

I can see two solutions.

1. Drop the flush entirely.  I'm not sure what it's intending to be doing, but the original fflush dates way back to pre-htslib when all this code was in io_lib.

2. Add a few hflush equivalent which is flush only, without the sync bit.

My instinct is to do 1.

**Comment by jmarshall** [comment]
`fflush()` just flushes the libc buffers out to the file descriptor, in case you have the same file open in several `FILE *`s. Or so sayeth the Linux man pages:

> Note  that fflush() only flushes the user-space buffers provided by the C library.  To ensure that the data is physically stored on disk the kernel buffers must be flushed too, for example, with sync(2) or fsync(2).

hFILE doesn't really believe in opening the same file more than once, especially for writing, so it doesn't have an equivalent of _flush-the-hFILE-buffers-only_; its `hflush()` has the meaning of _I want the bits on disk_. Hence what you're seeing.

So I'd agree with your option (1); with hFILE you mostly don't want to be explicitly flushing.

I feel like we've had this discussion before about some other (BAM?) code, and come to the conclusion of _don't call `hflush` so much then_. However see also f72add4a2c7dc75460a8558ebc41faeb99888fa2.

**Comment by jkbonfield** [comment]
Yes I found that bgzf change when I was looking back into this.  There it makes sense because we're doing it asynchronously and it's in order to prevent a potential hang at the end.

This I think is probably just accidental, and a rather embarrassingly long time between the change and noticing it.  I'll put that down to it needing a modern OS... it takes us an equally embarrassingly long time to install one ;-)
